django_celery_beat/decorators.py

```diff
@@ -0,0 +1,102 @@
+
+_app: Celery = app_or_default()
+
+# A list of periodic tasks that are to be connected when Celery is ready.
+# The tasks are stored as a list of (*arg, **kwarg) tuples.
+_periodic_tasks = []
+
+
+def _add_periodic_task(*args, **kwargs):
+    """
+    Queues a periodic task to be registered after Celery has finished initializing,
+    or registers it immediately if Celery is ready.
+    """
+    global _periodic_tasks
+
+    # No need to queue tasks, they can be added immediately.
+    if _app.configured:
+        _register_periodic_task(*args, **kwargs)
+    else:
+        # Register the signal callback the first time a task is queued.
+        if not _periodic_tasks:
+            _app.on_after_configure.connect(_register_all_periodic_tasks)
+
+        _periodic_tasks.append((args, kwargs))
+
+
+def _register_all_periodic_tasks(*args, **kwargs):
+    """
+    Registers each task that was queued by _add_periodic_task. While it would be
+    convenient to just do this directly in the `periodic_task` decorator, the problem
+    there is that the signal callback is stored on the stack and becomes dead as soon
+    as the decorator exits.
+    """
+    global _periodic_tasks
+
+    # Add each task.
+    for task in _periodic_tasks:
+        _register_periodic_task(*task[0], **task[1])
+
+    _periodic_tasks.clear()
+
+
+def _register_periodic_task(*task_args, **task_kwargs):
+    """Registers a single periodic task."""
+    _app.add_periodic_task(*task_args, **task_kwargs)
+
+
+def periodic_task(run_every, **task_kwargs):
+    """
+    Decorator for creating a periodic task.
+
+    `run_every` specifies when or how often the periodic task will be scheduled to run.
+
+    It supports several different types:
+
+    - `float`: interpreted as seconds.
+    - `timedelta`: interpreted as a regular time interval.
+    - `celery.schedules.crontab`: interpreted as an interval using crontab notation.
+    - `celery.schedules.solar`: interpreted as an interval based on solar occurences.
+
+    ### Example
+
+    ```
+        from django_celery_beat.decorators import periodic_task
+        from datetime import timedelta
+
+        @periodic_task(run_every=timedelta(minutes=5))
+        def say_hello():
+            print("Hello, world!")
+    ```
+
+    ### Resources
+
+    Info on crontab scheduling:
+
+    https://docs.celeryproject.org/en/v4.1.0/userguide/periodic-tasks.html#crontab-schedules
+
+    Info on solar scheduling:
+
+    https://docs.celeryproject.org/en/v4.1.0/userguide/periodic-tasks.html#solar-schedules
+    """
+
+    def wrapper(task_func):
+        # Wrap the decorated function to convert it into a celery task while also
+        # preserving its original properties so that a celery worker can find it.
+        @_app.task
+        @functools.wraps(task_func)
+        def wrapped_task(*args, **kwargs):
+            return task_func(*args, **kwargs)
+
+        _add_periodic_task(run_every, wrapped_task, **task_kwargs)
```

Comment:
Should there be further testing to demonstrate how the combination of `task_kwargs` + `args` + `kwargs` maps to the less flexible surface of [add_periodic_task](https://docs.celeryproject.org/en/latest/_modules/celery/app/base.html#Celery.add_periodic_task)?

Reply:
I added a bit more assertions for the kwargs. I'm not really 100% sure what you are looking to test specifically, let me know if you have a test case in mind

Reply:
There was more thinking out loud to this comment than was likely helpful. I'm biasing towards our older django-celery code here, but the most important one feels like `name=` since that's the value which could be used to access the `PeriodicTask` model, for us. 

So if you'd change `kwarg1` to name, and we make sure `add_periodic_task` got it, I'd be 😄 And thank you.